docs/proposals/check-design.md

```diff
@@ -0,0 +1,114 @@
+✓ Checked [succeeded]  application/vnd.oci.image.config.v1+json                                             581/581  B 100.00%  463ms
+  └─ sha256:aded1e1a5b3705116fa0a92ba074a5e0b0031647d9c315983ccba2ee5428ec8b
+✓ Checked [succeeded]  application/vnd.oci.image.manifest.v1+json                                         1022/1022  B 100.00%  437ms
+  └─ sha256:1c4eef651f65e2f7daee7ee785882ac164b02b78fb74503052a26dc061c90474
+Checked [failed]    [registry] localhost:5000/hello:v1
+
+Checked localhost:5000/hello:v1 in 15s. 1 check failed.
+[Failed]
+Error: oras check failed on f18232174bc9 application/vnd.oci.image.layer.v1.tar+gzip: layer size mismatch: expect 257, got 233
+```
+### Error messages:
+
+The `oras check` command should return refined error messages for failed checks. For example:
+```
+[Failed]
+Error: oras check failed on f18232174bc9 application/vnd.oci.image.layer.v1.tar+gzip: layer size mismatch: expect 257, got 233
+```
+
+
+## Validation behaviors (For manifest only, not index yet)
+
+When validating an image, the `oras check` command verifies the integrity of the manifest and all associated blobs. If the manifest includes a subject field, the referenced subject image is also validated recursively using the same process. The check proceeds through all components, reporting any failures, but does not stop at any failed check; instead, it continues until every component has been checked.
+
+* Step 1: Resolve the reference given by the user
+
+Returns an error if the reference fails to resolve, or resolves to a blob instead of a manifest.
+
+* Step 2: Validate the manifest against the resolved descriptor
+
+Fetch the manifest content by descriptor, and check if the **MediaType**, **Size** and **Digest** are consistent with the descriptor.
+
+* Step 3: Validate the blobs (config and layers) against the descriptors in the manifest
+
+Fetch the blobs (config and layers) by descriptors in the manifest, and check if the **Size** and **Digest** are consistent with the descriptors.
+
+
+```mermaid
+graph TD;
+    A[referrer of v1] -- subject --> B[manifest v1]
+```
+
+* Step 4: If the subject field is not empty, validate the subject manifest
+
+(Note: this is successor validation, enabled by default)
+
+Do step 2 with the descriptor in the subject field. Then do step 3.
+
+* Step 5: If --include-referrers flag is used, validate the referrers
+
+(Note: this is predecessor validation, only enabled by flag)
+
+Resolve the referrer descriptors, then do step 2 and 3.
+
+Verify that the subject descriptor of the referrers is consistent with the manifest.
+
+## PoC implementation
+https://github.com/oras-project/oras/pull/1801
+
+## Questions for discussion
+1. Should we support reference by digest?
+2. Should we support validating index? What should the behavior be?
+3. Should we support validating an entire repository? (`oras check localhost:5000/hello`, no image reference is given)
```

Comment:
We may want to think which use case that users want to check the entire repo?

Reply:
Two scenarios I can think of:

1. When a user fails to pull an image from the registry, he can run `oras check` to see if the image is corrupted. This helps the user to troubleshoot faster. 
2. User may want to set up a periodic pipeline (say once a month) to check that all current images in the production repository are intact.